geomesa-features/geomesa-feature-common/src/main/scala/org/locationtech/geomesa/features/ScalaSimpleFeature.scala

```diff
@@ -116,142 +139,180 @@ object ScalaSimpleFeature {
   }
 
   /**
     * Lazily evaluated, mutable simple feature implementation
     *
     * @param sft simple feature type
-    * @param initialId simple feature id
-    * @param reader lazily read attributes, must be thread-safe and already of the appropriate types
-    * @param userDataReader lazily read user data, must be thread-safe
     */
-  class LazyMutableSimpleFeature(
-      sft: SimpleFeatureType,
-      initialId: String,
-      private var reader: LazyAttributeReader,
-      private var userDataReader: LazyUserDataReader
-    ) extends AbstractMutableSimpleFeature(sft, initialId, null) {
+  class LazyMutableSimpleFeature private (sft: SimpleFeatureType)
+      extends AbstractMutableSimpleFeature(sft) with LazySimpleFeature with Sizable {
 
-    // we synchronize on the low-level words, in order to minimize contention
-    private val bits = WordBitSet(sft.getAttributeCount)
-    private lazy val attributes = Array.ofDim[AnyRef](sft.getAttributeCount)
-
-    private var phaser: Phaser = new Phaser(sft.getAttributeCount) {
-      override def onAdvance(phase: Int, registeredParties: Int): Boolean = {
-        // once all attributes have been read, dereference any backing resources so that they can be gc'd
-        reader = null
-        phaser = null
-        true
-      }
-    }
-
-    override lazy val getUserData: java.util.Map[AnyRef, AnyRef] = {
-      val ud = userDataReader.read()
-      userDataReader = null // dereference any backing resources
-      ud
+    /**
+     * Constructor
+     *
+     * @param sft simple feature type
+     * @param id simple feature id
+     * @param reader lazily read attributes, must be thread-safe and already of the appropriate types
+     * @param userDataReader lazily read user data, must be thread-safe
+     *
+     */
+    def this(sft: SimpleFeatureType, id: String, reader: LazyAttributeReader, userDataReader: LazyUserDataReader) = {
+      this(sft)
+      this.id = id
+      this.reader = reader
+      this.userDataReader = userDataReader
     }
 
     override def setAttributeNoConvert(index: Int, value: AnyRef): Unit = {
       if (reader != null) {
         val word = bits.word(index)
         word.synchronized {
           if (word.add(index)) {
-            phaser.arriveAndDeregister()
+            bits.synchronized {
+              if (attributes == null) {
+                attributes = Array.ofDim[AnyRef](sft.getAttributeCount)
+              }
+            }
+            if (count.decrementAndGet() == 0) {
+              // once all attributes have been read, dereference any backing resources so that they can be gc'd
+              reader = null
+              count = null
+            }
           }
         }
       }
       attributes(index) = value
     }
 
+    override def calculateSizeOf(): Long =
+      Sizable.sizeOf(this) + Sizable.deepSizeOf(id, bits, attributes, userData, count, reader, userDataReader)
```

Comment:
What order do we expect calculateSizeOf and setAttributeNoConverter to be called in?

I'm wondering if we should be counting the size of things like the reader and count which could null or used?

Reply:
i think they get nulled out after they're no longer needed - the deepSizeOf method will skip over null values, so they won't get counted if they are null.

Reply:
but to answer your more general question, they could be called in any order. in the standard query workflow, `setAttributeNoConvert` won't be called at all, and sizeOf will only be called if we have to sort on disk.